### PR TITLE
New shielder smart contract

### PR DESCRIPTION
**Pull request by kroist**
# Description

Initial implementation of shielder smart contract, consistent with https://docs.alephzero.org/aleph-zero/shielder/notes-and-accounts

Current public operations include deposit and withdrawal of PSP22 token

Zk relations are checked in module `mocked_zk.rs`

### CODE REVIEW

File: shielder/contract/lib.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {
+                nullifier_set: Mapping::new(),
+                notes: MerkleTree::new()
+            }
+        }
+
+        #[ink(message)]
+        pub fn add_note(&mut self, op_pub: OpPub, h_note_new: Scalar) -> Result<(), ShielderError>{
```
Comment (DamianStraszak) [inline shielder/contract/lib.rs:42]: If this is just about adding a new, fresh note, then it shouldn't take `op_pub` as input. It's just initializing the account then.

File: shielder/contract/lib.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            &mut self,
+            op_pub: OpPub,
+            h_note_new: Scalar,
+            merkle_root: Scalar,
+            nullifier_old: Scalar,
+            proof: ZkProof,
+        ) -> Result<(), ShielderError> {
+            self.process_operation(op_pub)?;
+            self.notes.is_historical_root(merkle_root)?;
+            (!self.nullifier_set
```
Comment (DamianStraszak) [inline shielder/contract/lib.rs:59]: I think it would be cleanest to have a separate function `fn nullify(&mut self, x: Scalar) -> Result<(), ...>` that adds to the nullifier or outputs error, if the element is there already. This way you can replace this, and also remove the line `self.nullifier_set.insert(nullifier_old, &());`.

File: shielder/contract/merkle.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+pub const DEPTH: usize = 10;
+
+#[ink::storage_item]
+#[derive(Debug)]
+pub struct MerkleTree {
+    nodes: Mapping<u32, Scalar>,
+    roots_log: Set<Scalar>,
+    next_leaf_id: u32,
+    sz: u32,
```
Comment (DamianStraszak) [inline shielder/contract/merkle.rs:16]: better to spell it out: `size` :)

Comment (DamianStraszak) [reply shielder/contract/merkle.rs:16]: Also please add a comment on what is this exactly

File: shielder/contract/mocked_zk.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+trait Hashable {
+    fn hash(&self) -> Scalar;
+}
+
+#[derive(Clone, Copy, scale::Encode, scale::Decode)]
+#[cfg_attr(
+    feature = "std", 
+    derive(scale_info::TypeInfo)
+)]
+
```
Comment (DamianStraszak) [inline shielder/contract/mocked_zk.rs:17]: You can remove the blank line here and in other such places.

File: shielder/contract/mocked_zk.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+struct Account {
+    balance_aleph: Scalar,
+    balance_usdt: Scalar,
+}
+
+impl Hashable for Account {
+    fn hash(&self) -> Scalar {
+        merkle::compute_hash(
+            self.balance_aleph,
+
```
Comment (DamianStraszak) [inline shielder/contract/mocked_zk.rs:79]: blank line, 
if you haven't please run `cargo fmt` (also `cargo clippy`)

File: shielder/contract/mocked_zk.rs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+const USDT_TOKEN: [u8; 32] = [0x2_u8; 32];
+
+impl Account {
+    // TODO: increase and decrease balances
+    fn update(&self, operation: Operation) -> Self {
+        match operation.op_pub {
+            OpPub::Deposit(amount, token, _) => {
+                let balance_usdt = self.balance_usdt;
+                if token.as_ref() == USDT_TOKEN {
+                    // decrease scalar by amount
```
Comment (DamianStraszak) [inline shielder/contract/mocked_zk.rs:94]: You can just increase here, right? Or is there a reason this is still TODO?

File: shielder/contract/mocked_zk.rs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+
+#[derive(Clone, Copy, scale::Encode, scale::Decode)]
+#[cfg_attr(
+    feature = "std", 
+    derive(scale_info::TypeInfo)
+)]
+
+pub struct ZkProof {
```
Comment (DamianStraszak) [inline shielder/contract/mocked_zk.rs:122]: Nice, you can write a comment that this is just a "mock proof", so opposite of zero-knowledge. But it's good :)

**Review by DamianStraszak — CHANGES_REQUESTED** [review]
Great piece of work! Some small comments only

File: shielder/contract/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+[package]
+name = "contract"
+version = "0.1.0"
+authors = ["[your_name] <[your_email]>"]
```
[the remark this answers is not in the document]
Comment (DamianStraszak) [reply shielder/contract/Cargo.toml:4]: I mean authors, homepage, and `name = "shielder"`

File: shielder/contract/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+edition = "2021"
+
+[dependencies]
+ink = { version = "4.2.0", default-features = false }
```
[the remark this answers is not in the document]
Comment (deuszx) [reply shielder/contract/Cargo.toml:8]: Actually, even 5.0-rc is out. @pmikolajczyk41 used it in his halo2-rsa-circuit https://github.com/Cardinal-Cryptography/rsa-challenge-with-halo2/blob/main/rsa_contract/Cargo.toml#L16 

It's indeed RC but likely to be released soon. It has some nice features but here, most useful one would most probably be ability to use drink! as the testing backend - https://github.com/Cardinal-Cryptography/rsa-challenge-with-halo2/blob/main/rsa_contract/src/tests.rs#L51

This is kinda being future-proof but this migration would likely happen anyway.

[the remark this answers is not in the document]
Comment (DamianStraszak) [reply shielder/contract/Cargo.toml:8]: Yes. @deuszx is right here. If we can go with 5.0 already then it's a huge win. I'm not experienced with drink! but it's the future for sure :)

[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply shielder/contract/Cargo.toml:8]: drink! already works with ink 5

File: shielder/contract/mocked_zk.rs (hunk 5 above)
Comment (Yithis) [reply shielder/contract/mocked_zk.rs:79]: If you are using VS code by any chance, it is possible to run at least first of them when saving. Just wanted to add this info as this is imo enormously helpful

File: shielder/contract/psp22.rs [hunk 10]
```diff
@@ -0,0 +1,120 @@
+use ink::{primitives::AccountId,prelude::vec::Vec};
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply shielder/contract/psp22.rs:1]: can't we depend on the actual psp22 crate?

File: shielder/contract/errors.rs [hunk 11]
```diff
@@ -0,0 +1,22 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
```
Comment (Yithis) [inline shielder/contract/errors.rs:1]: This is redundant as `lib.rs` is already affecting the whole crate`

Comment (Yithis) [reply shielder/contract/errors.rs:1]: Same comment apply to other files as well

File: shielder/contract/lib.rs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    use crate::{merkle::MerkleTree, psp22::PSP22, types::{Set, Scalar}, errors::ShielderError, mocked_zk::{ZkProof, self}};
+
+    #[derive(Clone, Copy, scale::Encode, scale::Decode)]
+    #[cfg_attr(
+        feature = "std", 
+        derive(scale_info::TypeInfo)
+    )]
+    pub enum OpPub {
+        Deposit(u128, AccountId, AccountId),
```
[the remark this answers is not in the document]
Comment (deuszx) [reply shielder/contract/lib.rs:22]: Or alternatively
```rust
pub enum OpPub {
  Deposit {
    amount: u128,
    account: AccountId, 
    ...
  }
}
```

File: shielder/contract/psp22.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (deuszx) [reply shielder/contract/psp22.rs:1]: You can look for inspiration here https://github.com/Cardinal-Cryptography/common-amm/blob/main/amm/contracts/pair/Cargo.toml#L17

File: shielder/contract/rust-toolchain [hunk 13]
```diff
@@ -0,0 +1 @@
+stable
```
[the remark this answers is not in the document]
Comment (Yithis) [reply shielder/contract/rust-toolchain:1]: I just notice that there is `rust-toolchain`  in root dir, so we should probably update that one to much what @DamianStraszak wrote about

File: shielder/contract/Cargo.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/Cargo.toml]: fixed

File: shielder/contract/types.rs [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+                .as_slice()
+                .try_into()
+                .unwrap(),
+        }
+    }
+}
+
+impl From<Scalar> for u128 {
+    fn from(value: Scalar) -> Self {
+        u128::from_le_bytes(value.bytes[0..16].try_into().unwrap())
```
Comment (pmikolajczyk41) [inline shielder/contract/types.rs:28]: in case anything changes, it'd be good to have a few test cases covered for this conversion

File: shielder/contract/mocked_zk/mod.rs [hunk 15]
```diff
@@ -0,0 +1,9 @@
+mod account;
+mod note;
+mod ops;
+pub mod relations;
+#[cfg(test)]
+mod tests;
+mod traits;
+
```
Comment (pmikolajczyk41) [inline shielder/contract/mocked_zk/mod.rs:8]: can you remove `proof.rs` file?

**Review by pmikolajczyk41 — COMMENTED** [review]
a few nits to the contract part, but in general a huge amount of a very nice code!

File: shielder/contract/Cargo.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/Cargo.toml]: added details, name is shielder-contract

File: shielder/contract/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+psp22 = { version = "0.2.1", default-features = false }
+
+scale = { package = "parity-scale-codec", version = "3", default-features = false, features = ["derive"] }
+scale-info = { version = "2.6", default-features = false, features = ["derive"], optional = true }
+
+[dev-dependencies]
+ink_e2e = "4.2.0"
```
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/Cargo.toml]: at the moment removed this dependency

File: shielder/contract/Cargo.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/Cargo.toml]: changed the version to 5.0-rc

File: shielder/contract/psp22.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/psp22.rs]: as I changed the ink version to 5.0.0-rc, no way to integrate the crate directly. created separate module traits

File: shielder/contract/merkle.rs (hunk 3 above)
Comment (kroist) [reply shielder/contract/merkle.rs]: changed to size and commented merkle

File: shielder/contract/mocked_zk.rs (hunk 5 above)
Comment (kroist) [reply shielder/contract/mocked_zk.rs]: now using formatting, later would add it in CI

File: shielder/contract/mocked_zk.rs (hunk 6 above)
Comment (kroist) [reply shielder/contract/mocked_zk.rs]: yes, wrote the whole module for mocked_zk

File: shielder/contract/errors.rs (hunk 11 above)
Comment (kroist) [reply shielder/contract/errors.rs]: removed it

File: shielder/contract/Cargo.toml [hunk 17]
```diff
@@ -0,0 +1,31 @@
+[package]
+name = "shielder"
```
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/Cargo.toml]: changed the name

File: shielder/contract/Cargo.toml [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+version = "0.1.0"
+authors = ["Cardinal"]
+homepage = "https://alephzero.org"
+repository = "https://github.com/Cardinal-Cryptography/aleph-node"
+edition = "2021"
+
+[dependencies]
+ink = { version = "4.3.0", default-features = false }
```
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/Cargo.toml]: changed the version

File: shielder/contract/lib.rs [hunk 19]
```diff
@@ -0,0 +1,112 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
```
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/lib.rs:6]: now using it

File: shielder/contract/lib.rs [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+                notes: MerkleTree::new(),
+            }
+        }
+
+        #[ink(message)]
+        pub fn add_note(
+            &mut self,
+            h_note_new: Scalar,
+            proof: ZkProof,
+        ) -> Result<(), ShielderError> {
```
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/lib.rs:71]: currently leaving it as todo

File: shielder/contract/lib.rs [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        fn nullify(&mut self, nullifier: Scalar) -> Result<(), ShielderError> {
+            if self.nullifier_set.contains(nullifier) {
+                Err(ShielderError::NullifierIsInSet)
+            } else {
+                self.nullifier_set.insert(nullifier, &());
+                Ok(())
+            }
+        }
```
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/lib.rs]: rewrote it

File: shielder/contract/lib.rs [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+        pub fn update_note(
+            &mut self,
+            op_pub: OpPub,
+            h_note_new: Scalar,
+            merkle_root: Scalar,
+            nullifier_old: Scalar,
+            proof: ZkProof,
+        ) -> Result<(), ShielderError> {
+            self.process_operation(op_pub)?;
+            self.notes.is_historical_root(merkle_root)?;
```
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/lib.rs:88]: reordered

File: shielder/contract/types.rs (hunk 14 above)
Comment (kroist) [reply shielder/contract/types.rs:28]: added tests for it

File: shielder/contract/mocked_zk/mod.rs (hunk 15 above)
Comment (kroist) [reply shielder/contract/mocked_zk/mod.rs:8]: removed

File: shielder/contract/Cargo.toml [hunk 23]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "shielder-contract"
+version = "0.1.0"
+authors = ["Cardinal"]
+homepage = "https://alephzero.org"
+repository = "https://github.com/Cardinal-Cryptography/aleph-node"
```
Comment (DamianStraszak) [inline shielder/contract/Cargo.toml:6]: repository should be `https://github.com/Cardinal-Cryptography/zk-apps/`

**Review by DamianStraszak — APPROVED** [review]
Very nice!

File: shielder/contract/Cargo.toml [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+edition = "2021"
+
+[dependencies]
+ink = { version = "4.3.0", default-features = false }
+psp22 = { version = "0.2.1", default-features = false }
+
+scale = { package = "parity-scale-codec", version = "3", default-features = false, features = ["derive"] }
+scale-info = { version = "2.6", default-features = false, features = ["derive"], optional = true }
```
[the remark this answers is not in the document]
Comment (kroist) [reply shielder/contract/Cargo.toml:13]: I was not able to compile the contract without scale deps, so I left them

File: shielder/contract/errors.rs [hunk 25]
```diff
@@ -0,0 +1,19 @@
+use crate::traits::psp22::PSP22Error;
+
+#[derive(PartialEq, Debug, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
```
Comment (pmikolajczyk41) [inline shielder/contract/errors.rs:4]: please use `#[ink::scale_derive(Encode, Decode, TypeInfo)]`; that way you don't need `scale` and `scale-info` dependencies

File: shielder/contract/merkle.rs [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+        Self {
+            nodes: Default::default(),
+            roots_log: Default::default(),
+            next_leaf_idx: 0,
+            size: (1 << DEPTH),
+        }
+    }
+}
+
+impl MerkleTree {
```
Comment (pmikolajczyk41) [inline shielder/contract/merkle.rs:44]: a todo (or better, a Jira item) - write tests for this module, especially for the edge cases

File: shielder/contract/traits/mod.rs [hunk 27]
```diff
@@ -0,0 +1 @@
+pub mod psp22;
```
Comment (pmikolajczyk41) [inline shielder/contract/traits/mod.rs:1]: why do we copy the whole PSP22 API instead of reusing psp22 crate? if it's the ink version conflict, I would rather move this code there - keeping it here is not a good idea (we pollute shielder code and might diverge from any development that is done in the reference repo)

**Review by pmikolajczyk41 — APPROVED** [review]

**Review by Yithis — APPROVED** [review]
